### PR TITLE
Flexible API for nearest neighbour search

### PR DESCRIPTION
**Pull request by anhldbk**
This is an implementation for issue [#527](https://github.com/piskvorky/gensim/issues/527)

I've added neighbor_sim.py in directory gensim/models in which I defined NeighborIndexer classes.
```python
class NeighborIndexer(object):
    """
    Base class for k-NN libraries
    """

    def __init__(self, model=None, **kwargs):
        """
        Create a new indexer
        :param model:   Instance of Doc2Vec or Word2Vec
        :param kwargs:  Additional named parameters
        """
        if model is None:
            raise Exception("Invalid model parameter. Please provide model with an instance of Doc2Vec or Word2Vec")
        if type(model) is Doc2Vec:
            self._init_doc2vec_(model)
        elif type(model) is Word2Vec:
            self._init_word2vec_(model)
        else:
            raise Exception("Invalid model parameter. Please provide model with an instance of Doc2Vec or Word2Vec")
        self.start_indexing()

    def _init_doc2vec_(self, model):
        docvecs = model.docvecs
        docvecs.init_sims()
        size = len(docvecs.doctag_syn0norm)

        for i in range(size):
            self.add_item(docvecs.offset2doctag[i], docvecs.doctag_syn0norm[i])

    def _init_word2vec_(self, model):
        # raise Exception("Not supported at the moment")
        model.init_sims()
        size = len(model.syn0norm)

        for i in range(size):
            self.add_item(model.index2word[i], model.syn0norm[i])

    def get_item(self, label):
        """
        Get an item's vector by its label
        :param label: The label
        :return: The item's vector if have. Otherwise, returns None
        """
        pass

    def add_item(self, label, vector):
        """
        Add an item to index
        :param label: the label of the item (must be unique).
        :param vector: the item's vector.
        """
        pass

    def start_indexing(self):
        """
        Start the indexing operation. This method is supposed to run after all items are added
        It may take time to finish.
        """
        pass

    def get_nearest_items(self, vector, top_n=10):
        """
        Get nearest items for an item described by its vector
        :param	vector: The vector
        :param	top_n: Number of nearest items to get
        :return	A list of tuples (label, similarity) for the nearest ones
        """
        pass

    def save(self, file_name):
        """
        Dump internal data into disk
        :param file_name:   Output file path
        :return: Returns true on success. Otherwise, returns false
        """
        pass

    def load(self, file_name):
        """
        Load previously dumped info
        :param file_name:  Input file paht
        :return: Returns true on success. Otherwise, returns false
        """
        pass
```
To make the integration process easier, I've modified most_similar() functions in doc2vec.py and word2vec.py by adding optional parameter *indexer*

For example: 
```python
# at line #409, doc2vec.py, the new function looks like this
def most_similar(self, positive=[], negative=[], topn=10, clip_start=0, clip_end=None, indexer=None):
```

To use the indexers, you may use the snippet below:
```python
from gensim.models.neighbor_sim import AnnoyIndexer
model = Doc2Vec.load(fname)

# create an indexer 
# features_num  = size of each document's vector
# You may need to optimize your *tree_size* parameter
indexer = AnnoyIndexer(model=model, features_num=100, tree_size=300)

# it may take time to finish the indexing operation

document = "Here comes the sun"
vector = model.features_num(document.lower().split())

# get most similar documents
similar_documents = model.most_similar([vector], indexer=indexer)
```

Currently, only Annoy-based NeighborIndexer is supported.

### CODE REVIEW

File: CHANGELOG.txt [hunk 1]
```diff
@@ -1,11 +1,7 @@
 Changes
 =======
 
-* Loading of pattern library in utils.py is only in lemmatize function (Jan Zikes, #461)
-  - utils.HAS_PATTERN, has also changed to utils.has_pattern()
-* Word2vec allows non-strict unicode error handling (ignore or replace). (Gordon Mohr, #466)
-
-0.12.3, 05/11/2015
```
Comment (piskvorky) [inline CHANGELOG.txt]: Why this change?

Comment (anhldbk) [inline CHANGELOG.txt]: @piskvorky Sorry, just an error during merging process. Pls discard it.

**Comment by piskvorky** [comment]
Thanks @anhldbk ! Will review asap :)

**Comment by piskvorky** [comment]
What's up with the commits by @tmylk in this PR?

**Comment by anhldbk** [comment]
@piskvorky Me? Actually I've just made my first pull request on github. lol It's ok now.

**Comment by tmylk** [comment]
@anhldbk Would be good to include this in the next release. For that purpose could you please:

- update CHANGELOG.txt with description of your changes
- add tests for regression of most_similar to ensure it didn't change. Existing tests either don't test the outcome [1](https://github.com/piskvorky/gensim/blob/dbe0b96c7935bd0916083a2f3fb5da63192bf569/gensim/test/test_doc2vec.py#L124), it is commented out [2](https://github.com/piskvorky/gensim/blob/839513f81e3aa42f490331fa80a28d13b7b7026f/gensim/test/test_word2vec.py#L189), or not robust [3](https://github.com/piskvorky/gensim/blob/839513f81e3aa42f490331fa80a28d13b7b7026f/gensim/test/test_word2vec.py#L357)
- resolve the merge conflicts in this pull request (follow github instructions)

**Comment by anhldbk** [comment]
@tmylk I'm glad to hear that. Actually it's the first time ever I've worked on a Github project. Would you pls tell me how to fulfill your request?

I'm thinking of these steps:
+ Pull the latest source from the main branch
+ Modify the source to include my logic.
+ Modify CHANGLOG.txt 
+ Run the tests --> How to run the tests? What will I do with the results?
+ Resolve any conflicts
+ Create a new pull request and close this one.

Is that ok ?

**Comment by tmylk** [comment]
@piskvorky ping

**Comment by anhldbk** [comment]
@tmylk So what have I got to do?

**Comment by piskvorky** [comment]
@anhldbk yes, your plan sounds good. I don't know why the commits are doubled in this PR -- creating a new pull request, branching off latest `develop`, should solve this.

File: gensim/models/neighbor_sim.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+from word2vec import Word2Vec
+
+from annoy import AnnoyIndex
+
+"""
+Using 3rd party k-NN libraries to improve performance for method DocvecsArray.most_similar().
+
+More information can be found at: https://github.com/piskvorky/gensim/wiki/Ideas-&-Feature-proposals#integrate-a-fast-k-nn-library
+
+Implemented by: Anh Le (anhldbk@gmail.com)
```
Comment (piskvorky) [inline gensim/models/neighbor_sim.py]: Belongs into the file header (see other source files in gensim).

File: gensim/models/neighbor_sim.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+Website: http://bigsonata.com
```
Comment (piskvorky) [inline gensim/models/neighbor_sim.py]: No such personal plugs in gensim.

File: gensim/models/neighbor_sim.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+
+Release 1: November 2015
```
Comment (piskvorky) [inline gensim/models/neighbor_sim.py]: Again, belongs either to the file header, or -- even better -- delete completely (history changes are handled by git).

File: gensim/models/neighbor_sim.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        Get an item's vector by its label
+        :param label: The label
+        :return: The item's vector if have. Otherwise, returns None
+        """
+        pass
+
+    def add_item(self, label, vector):
+        """
+        Add an item to index
+        :param label: the label of the item (must be unique).
```
Comment (piskvorky) [inline gensim/models/neighbor_sim.py]: We don't use Java-style docstrings in gensim. Either describe the params in free text, or structure using the [Google docstring style](https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments).

**Comment by anhldbk** [comment]
@piskvorky I'm gonna close this PR to make a new one.
